KNN - Iris

In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

df1=pd.read_csv("iris (1).csv")
df1.head()

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


In [6]:
X=df1.drop('species', axis=1)
Y=df1['species']

X_train, X_test, Y_train, Y_test=train_test_split(X, Y, test_size=0.2, random_state=42)

k=5
model=KNeighborsClassifier(n_neighbors=5)

model.fit(X_train, Y_train)

y_pred=model.predict(X_test)

accuracy=accuracy_score(Y_test, y_pred)

cm=confusion_matrix(Y_test, y_pred)

report=classification_report(Y_test, y_pred)

print("Accuracy: ", accuracy)
print("\nConfusion Matrix \n", cm)
print("\nClassification Report\n", report)

Accuracy:  1.0

Confusion Matrix 
 [[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]

Classification Report
               precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        10
  versicolor       1.00      1.00      1.00         9
   virginica       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



KNN - Diabetes

In [22]:
df2=pd.read_csv("diabetes.csv")
df2.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


Feature Scaling is required to bring all the attributes to the same range of values - as if they all have different ranges, the distance will be influenced by their extreme values


In [23]:
X=df2.drop('Outcome', axis=1)
Y=df2['Outcome']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

from sklearn.preprocessing import StandardScaler

scaler=StandardScaler()

X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)


In [24]:
model=KNeighborsClassifier(n_neighbors=5)

model.fit(X_train, Y_train)
y_pred=model.predict(X_test)

In [25]:
accuracy=accuracy_score(Y_test, y_pred)
cm=confusion_matrix(Y_test, y_pred)

print("Accuracy: ", accuracy)
print("Confusion Matrix: ", cm)

Accuracy:  0.6948051948051948
Confusion Matrix:  [[79 20]
 [27 28]]


StandardScaler -
Accuracy:  0.6948051948051948
Confusion Matrix:  [[79 20]
 [27 28]]

MinMaxScaler -
Accuracy:  0.6818181818181818
Confusion Matrix:  [[76 23]
 [26 29]]

KNN - Heart

In [27]:
df3=pd.read_csv("heart.csv")
df3.head()

df3.shape

(303, 14)

In [33]:
X=df3.drop('target', axis=1)
Y=df3['target']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

scaler=StandardScaler()

X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)

import numpy as np

#given that there are 303 rows, and even numbers create ties, our range for k is 1,3,5,7.....sqrt(303)
score=0;
res_k=0;
Y_pred=[];
for k in range (1, int(np.round(np.sqrt(df3.shape[0]))), 2) :
  model=KNeighborsClassifier(n_neighbors=k)
  model.fit(X_train, Y_train)
  y_pred=model.predict(X_test)
  accuracy=accuracy_score(Y_test, y_pred)

  if (accuracy > score):
    score=accuracy
    res_k=k
    Y_pred=y_pred

cm=confusion_matrix(Y_test, Y_pred)

report=classification_report(Y_test, Y_pred)

print("Accuracy: ", score)
print("Best K = ", res_k)
print("\nConfusion Matrix\n", cm)
print("\nClassification Report\n", report)




Accuracy:  0.9180327868852459
Best K =  7

Confusion Matrix
 [[27  2]
 [ 3 29]]

Classification Report
               precision    recall  f1-score   support

           0       0.90      0.93      0.92        29
           1       0.94      0.91      0.92        32

    accuracy                           0.92        61
   macro avg       0.92      0.92      0.92        61
weighted avg       0.92      0.92      0.92        61

